src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,385 @@
+/-
+Copyright (c) 2020 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import group_theory.congruence
+import linear_algebra.multilinear
+
+/-!
+# Tensor product of an indexed family of semimodules over commutative semirings
+
+We define the tensor product of an indexed family `s : ι → Type*` of semimodules over commutative
+semirings. We denote this space by `⨂[R] i, s i` and define it as `free_add_monoid (R × Π i, s i)`
+quotiented by the appropriate equivalence relation. The treatment follows very closely that of the
+binary tensor product in `linear_algebra/tensor_product.lean`.
+
+## Main definitions
+
+* `pi_tensor_product R s` with `R` a commutative semiring and `s : ι → Type*` is the tensor product
+  of all the `s i`'s. This is denoted by `⨂[R] i, s i`.
+* `tprod R f` with `f : Π i, s i` is the tensor product of the vectors `f i` over all `i : ι`.
+  This is bundled as a multilinear map from `Π i, s i` to `⨂[R] i, s i`.
+* `lift_add_hom` constructs an `add_monoid_hom` from `(⨂[R] i, s i)` to some space `F` from a
+  function `φ : (R × Π i, s i) → F` with the appropriate properties.
+* `lift φ` with `φ : multilinear_map R s E` is the corresponding linear map
+  `(⨂[R] i, s i) →ₗ[R] E`. This is bundled as a linear equivalence.
+
+## Notations
+
+* `⨂[R] i, s i` is defined as localized notation in locale `tensor_product`
+* `⨂ₜ[R] i, f i` with `f : Π i, f i` is defined globally as the tensor product of all the `f i`'s.
+
+## Implementation notes
+
+* We define it via `free_add_monoid (R × Π i, s i)` with the `R` representing a "hidden" tensor
+  factor, rather than `free_add_monoid (Π i, s i)` to ensure that, if `ι` is an empty type,
+  the space is isomorphic to the base ring `R`.
+* We have not restricted the index type `ι` to be a `fintype`, as nothing we do here strictly
+  requires it. However, problems may arise in the case where `ι` is infinite; use at your own
+  caution.
+
+## TODO
+
+* Define tensor powers, symmetric subspace, etc.
+* API for the various ways `ι` can be split into subsets; connect this with the binary
+  tensor product.
+* Include connection with holors.
+* Port more of the API from the binary tensor product over to this case.
+
+## Tags
+
+multilinear, tensor, tensor product
+-/
+
+noncomputable theory
+open_locale classical
+open function
+
+section semiring
+
+variables {ι : Type*} {R : Type*} [comm_semiring R]
+variables {R' : Type*} [comm_semiring R'] [algebra R' R]
```

Comment:
Requiring `algebra` feels quite strict here. Was `is_scalar_tower` not enough?

Reply:
I could probably do it with `[semimodule R' R] [is_scalar_tower R' R R]`, but isn't that just equivalent to algebra anyway?